KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+    }
+    function updateVoterAddress(address _addr) {
+        // ...
+        voterAddress = _addr;
+        stakingTracker.refreshVoter(address(this));
+    }
+}
+
+interface IStakingTracker {
+    /// @dev Re-evaluate Tracker contents related to the staking contract
+    /// @param staking  The CnStaking contract address
+    function refreshStake(address staking) external;
+
+    /// @dev Update a GC's voter address from given address
+    /// @param staking  The CnStaking contract address
+    function refreshVoter(address staking) external;
+}
+```
+
+#### Version identifier
+
+To distinguish the CnStakingContract and CnStakingV2 contracts, V2 will have VERSION value 2.
+
+```solidity
+abstract contract CnStakingV2 {
+    uint256 constant public VERSION = 2;
+}
+```
+
+### StakingTracker
+
+StakingTracker collects voting related data of each GC.
+
+- Staking amounts and voting powers \
+Stored separately in `Tracker` structs. Each `Tracker` struct can be updated between `trackStart` and `trackEnd` blocks, but becomes immutable afterwards, essentially freezing the voting powers. A `Tracker` struct is first created in `createTracker()`, and updated by `refreshStake()`
+
+- Each GC’s voter accounts \
+Stored in global mappings, and can be updated any time. Voter account mappings are updated by `refreshVoter()`
+
+#### Contract interface
+
+```solidity
+abstract contract StakingTracker {
+    struct Tracker {
+        // Tracker is only updated if trackStart <= block.number < trackEnd.
+        uint256 trackStart;
+        uint256 trackEnd;
+
+        // List of eligible nodes and their staking addresses.
+        // Initialized by createTracker().
+        address[] nodeIds;
+        mapping(address => address) rewardToNodeId;
+        mapping(address => address) stakingToNodeId;
+
+        // Balances and voting powers.
+        // Initialized by createTracker() and updated by refreshStake().
+        mapping(address => uint256) stakingBalances;
+        mapping(address => uint256) nodeBalances;
+        mapping(address => uint256) nodeVotes;
+        uint256 totalVotes;
+        uint256 eligibleNodes;
+    }
+
+    // Tracker objects. Added by createTracker().
+    mapping(uint256 => Tracker) private trackers; // trackerId => Tracker
+    uint256[] private allTrackerIds;
+
+    // 1-to-1 mapping between nodeId and voter account.
+    // Updated by refreshVoter().
+    mapping(address => address) private nodeIdToVoter; // nodeId => voterAddr
+    mapping(address => address) private voterToNodeId; // voterAddr => nodeId
+
+    /// @dev Creates a new Tracker and populate initial values from AddressBook
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external returns(uint256 trackerId);
+
+    /// @dev Re-evaluate Tracker contents related to the staking contract
+    /// Anyone can call this function, but `staking` must be a staking contract
+    /// registered to the AddressBook.
```

Comment:
@blukat29 What if `staking` is not a staking account? or what if it is a staking account but it is not connected to the address book?

Reply:
I'd like to let the implementation decide error handling policy.

As a side, what I would do is
- Reverts from `refreshStake()` call must not stop other functionality including stake withdrawl; Use `try-catch` or `.call()` to capture and silently ignore the revert.
- If CnStakingV2 is not registered to the AddressBook, then nothing gets written in `refreshStake()`.